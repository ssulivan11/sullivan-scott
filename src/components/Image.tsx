import React from 'react'

interface ImageProps {
  src: string
  alt: string
  width: number
  height: number
  quality: number
  templates: string[]
  lazy: boolean
  responsive: boolean
  className: string
  'data-test': string
}

const ImageSrcFormat = (src, width, height, quality, templates) => {
  const supportedResolutions = [1, 2]
  let srcSet = ''
  const templatesParam = templates.length ? `${templates.join('&')}&` : ``
  supportedResolutions.forEach((res) => {
    const widthParam = typeof width !== 'undefined' ? `w=${width * res}&` : ``
    const heightParam = typeof height !== 'undefined' ? `h=${height * res}&` : ``
    srcSet += `${src}?${templatesParam}${widthParam}${heightParam}qlt=${quality}&fmt=webp ${res}x, `
  })
  const widthParam = typeof width !== 'undefined' ? `w=${width}&` : ``
  const heightParam = typeof height !== 'undefined' ? `h=${height}&` : ``
  return {
    src: `${src}?${templatesParam}${widthParam}${heightParam}qlt=${quality}`,
    srcSet,
  }
}

const Image: React.FunctionComponent<ImageProps> = ({
  src,
  alt,
  width,
  height,
  quality = 75,
  templates = [],
  responsive = false,
  lazy = false,
  className,
  'data-test': dataTest = 'component-image',
  ...props
}) => {
  const { src: imageSrc, srcSet } = ImageSrcFormat(src, width, height, quality, templates)
  return (
    <picture>
      <source type='image/webp' srcSet={srcSet} />
      <source type='image/pjpeg' srcSet={imageSrc} />
      <img
        src={imageSrc}
        width={responsive ? '100%' : width}
        height={responsive ? '100%' : height}
        alt={alt}
        loading={lazy ? 'lazy' : 'auto'}
        data-test={dataTest}
        className={className}
        {...props}
      />
    </picture>
  )
}

export default Image
