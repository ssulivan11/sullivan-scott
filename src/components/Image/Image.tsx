import React from 'react'

interface ImageProps {
  src: string
  alt: string
  width: number
  height: number
  lazy?: boolean
  loading: boolean
  responsive?: boolean
}

const Image: React.FunctionComponent<ImageProps> = ({
  src,
  alt,
  width,
  height,
  responsive = false,
  lazy = true,
  ...props
}) => {
  return (
    <img
      src={src}
      width={responsive ? '100%' : width}
      height={responsive ? '100%' : height}
      alt={alt}
      loading={lazy ? 'lazy' : 'auto'}
      data-test='image-component'
      {...props}
    />
  )
}

export default Image
