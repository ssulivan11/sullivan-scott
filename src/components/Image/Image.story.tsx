import React from 'react'
import { withKnobs, number, text, boolean } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'

import Image from './Image'

const stories = storiesOf('Components | Image', module)

stories.addParameters({ options: { showPanel: true, isToolshown: true } })
stories.addDecorator(withKnobs)

function addKnobs(props) {
  return {
    ...props,
    src: text('src', props.src),
    alt: text('alt', props.alt),
    width: number('width', props.width),
    height: number('height', props.height),
    lazy: boolean('lazy', props.lazy),
    responsive: boolean('responsive', props.responsive),
  }
}

function getItem(props = {}) {
  const propsWithKnobs = addKnobs(props)
  return <Image {...propsWithKnobs} />
}

stories.add('Default', () =>
  getItem({
    src: 'https://ssullivan.info/images/icon-512x512.png',
    alt: 'scott sullivan logo',
    width: 200,
    height: 200,
    responsive: false,
  }),
)
