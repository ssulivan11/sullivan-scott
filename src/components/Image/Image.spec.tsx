import * as React from 'react'
import { shallow, mount } from 'enzyme'
import Image from './Image'

const defaultProps = {
  src: 'https://ssullivan.info/images/icon-512x512.png',
  alt: 'scott sullivan logo',
  width: 200,
  height: 200,
}

const responsiveProps = {
  ...defaultProps,
  responsive: true,
  lazy: false,
}

const setup = (props, renderer = shallow) => {
  const component = renderer(<Image {...props} />)

  return {
    getComponent: () => component,
    debug: () => component.debug(),
  }
}

describe('Image Component', () => {
  it('display default image', () => {
    const component = setup(defaultProps, mount)
    expect(component.getComponent()).toBeDefined()
  })

  it('display responsive container version', () => {
    const component = setup(responsiveProps, mount)
    expect(component.getComponent()).toMatchSnapshot()
  })
})
