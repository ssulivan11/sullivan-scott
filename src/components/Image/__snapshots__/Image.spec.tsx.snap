// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component display responsive container version 1`] = `ReactWrapper {}`;
